Index: src/main/java/org/cytoscape/view/vizmap/gui/internal/bypass/BypassManager.java
===================================================================
--- src/main/java/org/cytoscape/view/vizmap/gui/internal/bypass/BypassManager.java	(revision 25894)
+++ src/main/java/org/cytoscape/view/vizmap/gui/internal/bypass/BypassManager.java	(working copy)
@@ -27,7 +27,7 @@
 			.getLogger(BypassManager.class);
 
 	// TODO: how can we display compatible VP only?
-	private static final String DEF_RENDERER = "ding";
+	private static final String DEF_RENDERER = "wind";
 
 	// TODO: this should be a public string or enum managed in swing app.
 	private static final String MENU_KEY = "preferredMenu";
@@ -49,7 +49,7 @@
 	}
 
 	public void addBypass(RenderingEngineFactory<?> factory, Map props) {
-		if (props.containsValue("ding") == false)
+		if (props.containsValue("wind") == false)
 			return;
 
 		final VisualLexicon lexicon = factory.getVisualLexicon();
