Index: src/main/java/org/cytoscape/internal/view/NetworkViewManager.java
===================================================================
--- src/main/java/org/cytoscape/internal/view/NetworkViewManager.java	(revision 25942)
+++ src/main/java/org/cytoscape/internal/view/NetworkViewManager.java	(working copy)
@@ -124,7 +124,7 @@
 
 	// TODO: for now, use this as default. But in the future, we should provide
 	// UI to select presentation.
-	private static final String DEFAULT_PRESENTATION = "ding";
+	private static final String DEFAULT_PRESENTATION = "wind";
 
 	private final Map<CyTable, CyNetwork> nameTables;
 	private final CyNetworkViewManager networkViewManager;
